### PR TITLE
device: refactor device handles

### PR DESCRIPTION
**Pull request by gmarull**
This PR contains a series of patches that slightly refactor device handles:

- Renamed the feature to "device dependencies", since what was referenced as "handles" were dependencies in most cases. Device handles are a feature on their own.
- Make device dependencies optional, avoiding a second linker pass when not required (i.e., in the majority of cases)

### CODE REVIEW

File: kernel/Kconfig [hunk 1]
```diff
@@ -969,10 +969,18 @@ endmenu
...
+	  When enabled, device dependencies will be stored so that they can be
+	  queries at runtime. Device dependencies are typically inferred from
+	  devicetree.
+
+config DEVICE_DEPS_DYNAMIC
+	bool "Dynamic device dependencies"
+	depends on DEVICE_DEPS
 	help
-	  Hidden option that makes possible to manipulate device handles at
+	  Hidden option that makes possible to manipulate device dependencies at
```
Comment (nordicjm) [inline kernel/Kconfig:983]: Not hidden anymore

**Review by fabiobaltieri — DISMISSED** [review]

File: kernel/Kconfig [hunk 2]
```diff
@@ -969,10 +969,18 @@ endmenu
 
 menu "Device Options"
 
-config HAS_DYNAMIC_DEVICE_HANDLES
-	bool
+config DEVICE_DEPS
+	bool "Store device dependencies"
+	help
+	  When enabled, device dependencies will be stored so that they can be
+	  queries at runtime. Device dependencies are typically inferred from
```
Comment (nashif) [inline kernel/Kconfig:976]: queried

File: kernel/Kconfig [hunk 3]
```diff
@@ -969,10 +969,18 @@ endmenu
...
+	  devicetree.
+
+config DEVICE_DEPS_DYNAMIC
+	bool "Dynamic device dependencies"
+	depends on DEVICE_DEPS
 	help
-	  Hidden option that makes possible to manipulate device handles at
+	  Option that makes possible to manipulate device dependencies at
```
Comment (nashif) [inline kernel/Kconfig:983]: makes it possible....

File: CMakeLists.txt [hunk 4]
```diff
@@ -907,27 +907,27 @@ if(CONFIG_HAS_DTS)
...
     --start-symbol "$<TARGET_PROPERTY:linker,devices_start_symbol>"
     VERBATIM
     DEPENDS ${ZEPHYR_LINK_STAGE_EXECUTABLE}
     )
-  set_property(GLOBAL APPEND PROPERTY GENERATED_APP_SOURCE_FILES dev_handles.c)
+  set_property(GLOBAL APPEND PROPERTY GENERATED_APP_SOURCE_FILES device_deps.c)
 
-  # gen_handles runs on `__device_handles_pass1` so pass this info to the linker script generator
-  list(APPEND LINKER_PASS_${ZEPHYR_CURRENT_LINKER_PASS}_DEFINE "LINKER_DEVICE_HANDLES_PASS1")
+  # gen_deps runs on `__device_deps_pass1` so pass this info to the linker script generator
```
Comment (aescolar) [inline CMakeLists.txt:929]: gen_deps -> gen_device_deps

File: kernel/Kconfig [hunk 5]
```diff
@@ -969,10 +969,18 @@ endmenu
 
 menu "Device Options"
 
-config HAS_DYNAMIC_DEVICE_HANDLES
-	bool
+config DEVICE_DEPS
+	bool "Store device dependencies"
+	help
+	  When enabled, device dependencies will be stored so that they can be
```
Comment (aescolar) [inline kernel/Kconfig:975]: Would it nicer to be a bit more explicit saying that they will be stored in the image, and that disabling this feature can typically save from a few tenths to a couple hundredth bytes of ROM?

**Review by aescolar — COMMENTED** [review]
Thanks @gmarull 
Just a typo and a minor suggestion (It is up to you if you want to do anything about the suggestion or not)
Otherwise +1

File: CMakeLists.txt (hunk 4 above)
Comment (gmarull) [reply CMakeLists.txt]: adjusted

File: kernel/Kconfig (hunk 5 above)
Comment (gmarull) [reply kernel/Kconfig:975]: Added a comment about ROM (or RAM) usage

File: kernel/Kconfig (hunk 3 above)
Comment (gmarull) [reply kernel/Kconfig]: fixed

File: kernel/Kconfig (hunk 2 above)
Comment (gmarull) [reply kernel/Kconfig]: fixed

**Review by aescolar — APPROVED** [review]
+1

**Review by fabiobaltieri — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Pull request merged**